packages/lesswrong/components/users/NewUserCompleteProfile.tsx

```diff
@@ -0,0 +1,161 @@
+import { gql, useMutation } from "@apollo/client";
+import { Button, Checkbox, FormControlLabel, TextField, Typography } from "@material-ui/core";
+import React, { useState } from "react";
+import { forumTypeSetting, siteNameWithArticleSetting } from "../../lib/instanceSettings";
+import { Components, registerComponent } from "../../lib/vulcan-lib";
+import { useMessages } from "../common/withMessages";
+import { useCurrentUser } from "../common/withUser";
+
+const styles = (theme: ThemeType): JssStyles => ({
+  root: {
+    background: 'white',
+    padding: theme.spacing.unit * 6
+  },
+  title: {
+    marginTop: 0
+  },
+  section: {
+    marginTop: theme.spacing.unit * 6
+  },
+  sectionHelperText: {
+    color: theme.palette.grey[600],
+    fontStyle: 'italic',
+    fontSize: '1rem'
+  },
+  submitButtonSection: {
+    marginTop: theme.spacing.unit * 3
+  }
+});
+
+type NewUserCompleteProfileProps = {
+  classes: ClassesType
+}
+
+function prefillUsername(maybeUsername: string | undefined | null): string {
+  if (!maybeUsername) return ''
+  if (/^\S+@\S+\.\S+$/.test(maybeUsername)) return ''
+  if (/new_user_\d+/.test(maybeUsername)) return ''
+  return maybeUsername
+}
+
+const NewUserCompleteProfile: React.FC<NewUserCompleteProfileProps> = ({ classes }) => {
+  // TODO: Prefill with existing username if it's not an email / new_user_N
+  const currentUser = useCurrentUser()
+  const [username, setUsername] = useState(prefillUsername(currentUser?.displayName))
+  const [subscribeToDigest, setSubscribeToDigest] = useState(false)
+  const [validationError, setValidationError] = useState('')
+  const [updateUser] = useMutation(gql`
+    mutation NewUserCompleteProfile($username: String!, $subscribeToDigest: Boolean!) {
+      NewUserCompleteProfile(username: $username, subscribeToDigest: $subscribeToDigest) {
+        username
+        slug
+        displayName
+      }
+    }
+  `, {refetchQueries: ['getCurrentUser']})
```

Comment:
I think this works in principle, but merging fragments in Apollo has been a trouble-spot historically. If any troubles related to this come up in testing, I'd probably replace this with an `onFinish` that does a hard-refresh of the tab.

Reply:
I'm curious if you remember an example and could easily summarize or link to discussion of it?